### ISSUE TITLE
Why are many Stormpath error codes not covered by i18n.properties?

### ISSUE THREAD

**Issue by george-hawkins-aa** [stormpath/stormpath-sdk-java#903]
I just tried logging in with a deliberately invalid password.

This results in an error with the message key `stormpath.web.errors.7100`.

This does not have an entry in [`i18n.properties`](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/servlet/src/main/resources/com/stormpath/sdk/servlet/i18n.properties).

Providing at least the English versions of these errors was covered by #763 - which is now marked as closed.

If one looks at `i18n.properties` one sees many codes are covered, but some like 7100 are not. An invalid password must be amongst one of the most common errors seen so its omission seems odd.

7100 is covered by [`api-errors.properties`](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/servlet/src/test/resources/com/stormpath/sdk/servlet/api-errors.properties) under the test resources.

Here it has an overly specific value (which is fair enough for test) which gives away the fact that the problem is with the password:

    status.7100.dev = Login attempt failed because the specified password is incorrect.

So why despite all this does one still see "Invalid username or password." in the web frontend? The answer is that you hit the `NoSuchMessageException` catch block of `getMessage(String key, String defaultMessage, Locale locale, Object... args)` in [`AbstractStormpathWebMvcConfiguration`](https://github.com/stormpath/stormpath-sdk-java/blob/master/extensions/spring/stormpath-spring-webmvc/src/main/java/com/stormpath/spring/config/AbstractStormpathWebMvcConfiguration.java) which logs a warning and then returns a default value - in this case the value retrieved for `com.stormpath.sdk.servlet.mvc.LoginErrorModelFactory.INVALID_LOGIN_MESSAGE`.

So the end user sees something OK - but the logs get spammed with warnings and stack traces.

**Comment by dogeared** [comment]
Hi @george-hawkins-aa: Thanks for this! I've added this issue to our upcoming 1.0.3 release.

**Comment by dogeared** [comment]
Also, just to be clear: we don't ever want the dev messages in i18n.properties because we don't want the end-user seeing those messages. However, we can do a `log.debug` rather than having the logs fill up with exceptions.

**Comment by george-hawkins-aa** [comment]
We recently employed a pen-test company - one of the interesting attacks they demonstrated involved the original Stormpath login page.

We now have our own custom login page, so I'm not sure if this is still the case, but at one stage I think the logic when an error occurred was to _redirect_ the user like so:

    http://my.host/login?error=message.key

The Thymeleaf login page would then take the error key (which in this situation is echoed back to it via the client rather than coming directly from the backend) and then look it up.

If the key wasn't found it would simply use the key name itself as the text to be displayed. This meant an attacker could do something like this:

    http://my.host/login?error=

The pen-test company did this quite nicely - they could encode things so the HTML was way less obvious and their HTML could essentially comment out the real login form's HTML, replacing it with their own form (which obviously forwarded any login credentials to them).

As a fix we decided to disable the behavior of blindly echoing back unknown keys and so overrode the default message logic like so:

```Java
ResourceBundleMessageSource bean = new ResourceBundleMessageSource() {
    @Override
    protected String getDefaultMessage(String code) {
        logger.error("missing key - {}", code);
        return MISSING_KEY + HtmlUtils.htmlEscape(code);
    }
}
```

At the time of doing this I knew there was bound to be some logic that would depend on sometimes getting a null value back here. And that's exactly the case with the current Stormpath logic :(

It's actually higher up that the choice to use the key as the text is taken but this is the only point in the whole process at which there's an easy to get at an overrideable component in the whole message lookup process.

Also I'm unhappy with our above anyway solution as the `htmlEscape` encodes at a very low level (Spring message retrieval) the assumption that all messages ultimately end up being used for display in HTML (but I guess instead we could just strip everything except `[a-zA-Z0-9._]` which is probably fit for anything or omit the key value altogether from the return value, only including it in the log message).

PS I completely understand that the values in `api-errors.properties` are too specific for end user messages and give away too much information to a potential attacker.
